# (PART) Generation Proxy Selection Mapping {-}

# Introduction

The current project involved Generation Proxy Selection Mapping of swine and was the first of a series of projects completed by Caleb J. Grohmann as a part of his PhD program in Bioinformatics. Key collaborators on this project included Dr. Jared Decker (PhD advisor) and Dr. Tim Safranski (secondary advisor) from the University of Missouri - Columbia, and Dr. Clint Schwab (mentor/advisor), Dr. Caleb Shull (mentor/advisor), and Dr. Tamar Crum (data management/geneticist) from The Maschhoff's LLC. All relevant data files were uploaded to a shared Box folder (Dr. Decker, Caleb, and Dr. Crum) by Dr. Crum and downloaded to Caleb's local machine on September 9, 2020. All data analysis and communication of results was completed by Caleb J. Grohmann thereafter.

# Background

In swine breeding programs, the selection index has served as the primary method for assessing the aggregate genetic merit, for over 75 years, of boars and gilts by combining data from multiple traits that are a part of the broader breeding objective (Hazel et al., 1994). Economically relevant traits that are utilized in selection indices are generally quantitative in nature; thus, these traits are controlled by large numbers of genes, termed quantitative trait loci (**QTL**) by Geldermann (1975; Knol et al., 2016). The use of selection indices, i.e., artificial selection pressure, in pig breeding programs has been proven to cause significant changes to the mean phenotype of any one trait considered within the breeding objective. However, artificial selection pressure, especially over relatively short time scales, causes only subtle changes to allele frequencies at QTLs across the genome (Rowan et al., 2020). While allele frequencies change at QTLs for traits included in the selection index, neighboring loci to these QTLs can also "hitchhike" along with the primary locus under selection due to linkage disequilibrium (Decker et al., 2012). In addition, allele frequencies at loci that control traits that are not explicitly included in the selection index have been shown to undergo frequency changes as a result of artificial selection pressure applied in livestock breeding programs (Decker et al., 2012; Rowan et al., 2020).

There is much interest within the area of livestock genomics in deciphering the genetic basis of phenotypic diversity in species raised in animal agriculture for meat production (Qanbari and Simianer, 2014). Understanding selection signatures in livestock populations (i.e., allele frequency changes due to selection) is of paramount importance when evaluating the genomic basis of phenotypic variation within genetic line, breeds, or entire livestock populations over time. The identification of selection signatures assumes detection of areas of the genome with loci that have been subjected to a rapid increase in allele frequency, caused by forces of selection (Gurgul et al., 2018). With the advent of high-density genetic maps [i.e., single nucleotide polymorphism (**SNP**) arrays] and increased computing resources, statistical analysis of selection signatures is efficient. Identification of regions of the genome that have been altered due to artificial selection pressure is highly beneficial in ascertaining QTLs under selection. When results of selection signature analyses are combined with results from phenotype-based genome wide association analyses (**GWAAs**), QTLs that have contributed to phenotypic variation of traits within breeding objectives can be confirmed (Qanbari and Simianer, 2014). Moreover, there are opportunities within selection signature analyses to evaluate results within or across genetic lines or breeds, which can highlight differences in selection objectives across populations. Selection signature analyses are not only beneficial in increasing knowledge with respect to selection and evolution of species. For example, using results from selection signature analyses, SNP assays used for genomic prediction of breeding values in livestock populations can be refined in order to reduce extraneous statistical noise and increase prediction accuracy.

Generation Proxy Selection Mapping (**GPSM**) has been proposed as a powerful analytical method for detection of artificial selection signatures over relatively short time scales in breeds of cattle (Decker et al., 2012; Rowan et al., 2020). In this approach, birth date (or pedigree age) is fit as the dependent variable, and SNPs that are strongly associated or predictive of birth date are identified. If a SNP is under strong directional selection pressure, changes in allele frequency will generally be consistent over time, and and individual's genotype will be strongly predictive of birth date. Several studies have been conducted that have detected selection signals associated with growth and reproductive traits, coat color, and ear phenotype using more common statistical techniques within population genetics (Gurgul et al., 2018). However, a major advantage of GPSM methodology over other methods to elucidate loci under selection is the ability to adjust for confounding due to sampling bias (non-random ascertainment of genotype samples, population structure, or kinship) with the use of a genomic relationship matrix (**GRM**). In addition, estimation of genetic variance components (such as SNP heritability and genetic correlations) for birth date within and across breeds of pigs can provide comparison of artificial selection pressure across genetic lines that are more or less similar in terms of genotype. To date, there have been no large-scale studies that have utilized GPSM to map selection signatures in commercial swine populations. Thus, the objectives of the current study were to 1) use GPSM to identify loci under artificial selection in three purebred genetic lines and one crossbred genetic line of pigs and 2) compare and contrast the effect of artificial selection patterns among genotypes of each genetic line in the context of commerical pig production.
