# Results

```{r n-snps, echo=FALSE}
subgroups <- tibble(Subgroup_Number = c(1:12),
                    Subgroup = c("Crossbred", "Duroc", "Landrace", "Yorkshire",
                           "Crossbred and Duroc", "Crossbred and Landrace", "Crossbred and Yorkshire",
                           "Duroc and Landrace", "Duroc and Yorkshire", "Landrace and Yorkshire",
                           "Duroc, Landrace, and Yorkshire", "Crossbred, Duroc, Landrace, and Yorkshire"),
                    Subgroup_Code = c("c", "d", "l", "y", "cd", "cl", "cy", "dl", "dy", "ly", "dly", "cdly"),
                    Plot_Code = c("C", "D", "L", "Y", "C and D", "C and L", "C and Y", "D and L", "D and Y", "L and Y", "D, L, and Y", "C, D, L, and Y"),
       `Pigs_n` = c(8447, 16633, 18834, 17764, 25091, 27281, 26221, 35480, 34417, 36609, 53256, 61702),
       `SNPs_n` = c(46529, 38377, 45316, 45196, 46348, 46431, 46426, 46064, 46143, 46236, 46421, 46453))

subgroups$Subgroup <- factor(subgroups$Subgroup,
                             levels = c("Crossbred", "Duroc", "Landrace", "Yorkshire",
                           "Crossbred and Duroc", "Crossbred and Landrace", "Crossbred and Yorkshire",
                           "Duroc and Landrace", "Duroc and Yorkshire", "Landrace and Yorkshire",
                           "Duroc, Landrace, and Yorkshire", "Crossbred, Duroc, Landrace, and Yorkshire"))
subgroups$Subgroup_Code <- factor(subgroups$Subgroup_Code,
                             levels = c("c", "d", "l", "y", "cd", "cl", "cy", "dl", "dy", "ly", "dly", "cdly"))
subgroups$Plot_Code <- factor(subgroups$Plot_Code,
                             levels = c("C", "D", "L", "Y", "C and D", "C and L", "C and Y", "D and L", "D and Y", "L and Y", "D, L, and Y", "C, D, L, and Y"))
```

```{r pheno, echo=FALSE}
setwd("C:/Users/cgroh/OneDrive/Desktop/PhD/phd-bookdown/pheno")

file_list <- list.files(path="C:/Users/cgroh/OneDrive/Desktop/PhD/phd-bookdown/pheno")
pheno_qc <- tibble()

for (i in 1:length(file_list)){
  temp_data <- fread(file_list[i],
                     select = c(1, 3),
                     col.names = c("PigID", "Age")) %>%
    mutate(Subgroup_Code = file_list[i],
           Subgroup_Code = gsub("_uni.pheno", "", Subgroup_Code))
  pheno_qc <- rbindlist(list(pheno_qc, temp_data), use.names = T)
}

pheno_qc <- pheno_qc %>% 
  left_join(., subgroups[,1:4], by="Subgroup_Code")

rm(temp_data, file_list, i)
```
```{r line-breed, echo=FALSE}
ped_plus_intmd <- ped_plus_intmd %>%
  mutate(Line = ifelse(Line == 10, "Landrace",
                       ifelse(Line == 11, "Yorkshire",
                              ifelse(Line == 1006, "Duroc", "Crossbred"))))
```

## Exploratory Data Analysis

### Descriptive Statistics

```{r table-5-1}
pheno_qc %>%
  group_by(Subgroup_Number, Subgroup) %>%
  summarise(N = n(),
            Mean = round(mean(Age), 1),
            Median = median(Age),
            SD = round(sd(Age), 2),
            Minimum = min(Age),
            Maximum = max(Age)) %>%
  rename(`Subgroup` = Subgroup_Number,
         `Genetic lines` = Subgroup) %>%
  kable(.,
        row.names = FALSE,
        align = 'c',
        caption = "Descriptive statistics AGE by subgroup after genotype quality control.")
```

### Raw Distributions

#### One Breed

```{r figure-5-1, out.width='90%', fig.align='center', fig.cap='Histograms of AGE by one-breed subgroups.'}
ggplot(filter(pheno_qc, Subgroup_Code %in% c("c", "d", "l", "y")), aes(x=Age)) + 
  geom_histogram(color="white", fill="steelblue1", position = "dodge") +
  facet_wrap(~Subgroup, nrow = 2) +
  scale_x_continuous(expand = c(0,3)) +
  scale_y_continuous(expand = c(0,0), limits = c(0, 4000)) +
  geom_hline(yintercept = 0) +
  labs(y = "No. of pigs", x = "Age, months") +
  theme_classic() +
  theme(strip.background = element_rect(fill="grey70")) +
  theme(strip.text = element_text(color = 'white', face = "bold", size = "12"),
        axis.title = element_text(face = "bold"))
```

#### Two Breeds

```{r figure-5-2, out.width='90%', fig.align='center', fig.cap='Histograms of AGE by two-breed subgroups.'}
ggplot(filter(pheno_qc, Subgroup_Code %in% c("cd", "cl", "cy", "dl", "dy", "ly")), aes(x=Age)) + 
  geom_histogram(color="white", fill="steelblue1", position = "dodge") +
  facet_wrap(~Subgroup, nrow = 3) +
  scale_x_continuous(expand = c(0,3)) +
  scale_y_continuous(expand = c(0,0), limits = c(0, 4000)) +
  geom_hline(yintercept = 0) +
  labs(y = "No. of pigs", x = "Age, months") +
  theme_classic() +
  theme(strip.background = element_rect(fill="grey70")) +
  theme(strip.text = element_text(color = 'white', face = "bold", size = "12"),
        axis.title = element_text(face = "bold"))
```

#### Three+ Breeds

```{r figure-5-3, out.width='90%', fig.align='center', fig.cap='Histograms of AGE by subgroups of three or more breeds.'}
ggplot(filter(pheno_qc, Subgroup_Code %in% c("dly", "cdly")), aes(x=Age)) + 
  geom_histogram(color="white", fill="steelblue1", position = "dodge") +
  facet_wrap(~Plot_Code, nrow = 1) +
  scale_x_continuous(expand = c(0,3)) +
  scale_y_continuous(expand = c(0,0), limits = c(0, 5000)) +
  geom_hline(yintercept = 0) +
  labs(y = "No. of pigs", x = "Age, months") +
  theme_classic() +
  theme(strip.background = element_rect(fill="grey70")) +
  theme(strip.text = element_text(color = 'white', face = "bold", size = "12"),
        axis.title = element_text(face = "bold"))
```

## Univariate Variance Component Estimation

```{r figure-5-4, out.width='90%', fig.align='center', fig.cap='Proportion of variance in AGE explained by SNPs across various subgroups and subsets of SNPs.'}
uvce <- read.csv("results/uvce.csv",
                 header = TRUE)
uvce$Breeds_N <- factor(uvce$Breeds_N,
                        levels = c("One Breed", "Two Breeds", "Three+ Breeds"))
uvce$Plot_Code <- factor(uvce$Plot_Code,
                             levels = c("C", "D", "L", "Y", "C and D", "C and L", "C and Y", "D and L", "D and Y", "L and Y", "D, L, and Y", "C, D, L, and Y"))

library(RColorBrewer)
ggplot(uvce, aes(x = Plot_Code, y = PVE, color = SNP_Group)) +
  geom_point(aes(size = Pigs_n)) +
  scale_y_continuous(limits = c(0,1), breaks = round(seq(0, 1, by = 0.1), 2), expand = c(0,0)) +
  labs(x = "Subgroup", y = "Proportion of Variance Explained by SNPs",
       color = "GRM", size = "Sample Size") +
  theme_light() +
  theme(axis.text.x = element_text(angle = 45, vjust = .5),
        axis.title = element_text(face = "bold"),
        strip.background = element_rect(fill="grey70"),
        strip.text = element_text(color = 'white', face = "bold", size = "12"),
        legend.title = element_text(face = "bold")) +
  facet_grid(~Breeds_N, scales = "free") +
  scale_color_brewer(palette="RdBu")
```

## Bivariate Variance Component Estimation

```{r figure-5-5, out.width='90%', fig.align='center', fig.cap='Genetic correlation between breeds for AGE across various subgroups and subsets of SNPs.'}
bvce <- read.csv("results/bvce.csv",
                 header = TRUE)
bvce$Plot_Code <- factor(bvce$Plot_Code,
                             levels = c("C", "D", "L", "Y", "C and D", "C and L", "C and Y", "D and L", "D and Y", "L and Y", "D, L, and Y", "C, D, L, and Y"))

ggplot(bvce, aes(x = Plot_Code, y = rG, color = SNP_Group, shape = Breed)) +
  geom_point(size = 4) +
  scale_y_continuous(limits = c(-0.2,1), breaks = round(seq(-0.2, 1, by = 0.1), 2), expand = c(0,0)) +
  labs(x = "Subgroup", y = "Genetic correlation (rG)",
       color = "SNP Group", shape = "GRM") +
  theme_light() +
  theme(axis.text.x = element_text(angle = 45, vjust = .5),
        axis.title = element_text(face = "bold"),
        strip.background = element_rect(fill="grey70"),
        strip.text = element_text(color = 'white', face = "bold", size = "12"),
        legend.title = element_text(face = "bold")) +
  scale_color_brewer(palette="RdBu") +
  geom_hline(yintercept = 0, size = 1)
```

## Genome-wide Association Analyses

```{r mlma, echo=FALSE}
library(qvalue)
file_list <- list.files(path = "results/mlma")
mlma <- tibble()

setwd("results/mlma")

for (i in 1:length(file_list)){
  temp_data <- fread(file_list[i])
  q_obj <- qvalue(temp_data$p)
  temp_data$q <- q_obj$qvalue
  temp_data <- temp_data %>%
    mutate(Subgroup_Code = file_list[i],
           Subgroup_Code = gsub("_all_gwaa.mlma", "", Subgroup_Code))
  
  mlma <- rbindlist(list(mlma, temp_data), use.names = T)
}

mlma <- mlma %>% 
  left_join(., subgroups[c(1:6)], by="Subgroup_Code") 

mlma <- mlma %>%
  mutate(Subgroup_Number = ifelse(Subgroup_Code == "dgdrop", 2,
                                  ifelse(Subgroup_Code == "lgdrop", 3,
                                         ifelse(Subgroup_Code == "ygdrop", 4, Subgroup_Number))),
         Subgroup = ifelse(Subgroup_Code == "dgdrop", "Simulated Duroc",
                                  ifelse(Subgroup_Code == "lgdrop", "Simulated Landrace",
                                         ifelse(Subgroup_Code == "ygdrop", "Simulated Yorkshire", as.character(Subgroup)))),
         Plot_Code = ifelse(Subgroup_Code == "dgdrop", "Sim D",
                                  ifelse(Subgroup_Code == "lgdrop", "Sim L",
                                         ifelse(Subgroup_Code == "ygdrop", "Sim Y", as.character(Plot_Code)))),
         SNPs_n = ifelse(Subgroup_Code == "dgdrop", 38376,
                                  ifelse(Subgroup_Code == "lgdrop", 45324,
                                         ifelse(Subgroup_Code == "ygdrop", 45198, SNPs_n))),
         Pigs_n = ifelse(Subgroup_Code == "dgdrop", 16633,
                                  ifelse(Subgroup_Code == "lgdrop", 18834,
                                         ifelse(Subgroup_Code == "ygdrop", 17764, Pigs_n))))

rm(temp_data, file_list, i, q_obj)
```

### Frequency of Detected Associations

```{r table-5-2}
mlma %>%
  mutate(Threshold = 0.05,
         Adjusted_Threshold = Threshold/SNPs_n,
         p_sig = p < Adjusted_Threshold,
         q_sig = q < 0.10) %>%
  group_by(Subgroup_Number, Subgroup, Pigs_n, SNPs_n) %>%
  summarise(N_p = sum(p_sig),
            N_q = sum(q_sig)) %>%
  rename(`Subgroup` = Subgroup_Number,
         `Genetic lines` = Subgroup,
         `Pigs, n` = Pigs_n,
         `SNPs, n` = SNPs_n,
         `Signifigant SNPs (Bonferroni), n` = N_p,
         `Signifigant SNPs (Q-value), n` = N_q) %>%
  kable(.,
        row.names = FALSE,
        align = 'c',
        caption = "Number of statistically significant SNPs per subgroup based on Bonferroni (alpha/N, SNPs) and Q-value (Q < 0.10) thresholds.")
```

### Allele Substitution Effects

```{r figure-5-6, out.width="100%", fig.cap='Top allele substitution effects (beta) by one breed subgroups.', fig.align='center'}
top_betas <- mlma %>% 
  filter(Subgroup_Code %in% c("c", "d", "l", "y")) %>% 
  filter(q < 0.10) %>%
  group_by(Plot_Code) %>%
  slice_max(order_by = b, n = 10) %>%
  arrange(-b)

top_betas %>% 
  mutate(SNP = reorder(SNP, b)) %>%
  group_by(Subgroup, SNP) %>% 
  arrange(b) %>% 
  ungroup() %>% 
  mutate(SNP = factor(paste(SNP, Subgroup, sep = "__"), 
                       levels = rev(paste(SNP, Subgroup, sep = "__")))) %>%
ggplot(., aes(x = SNP, y = b)) +
  geom_errorbar(aes(ymin = b - se*1.96, ymax = b + se*1.96), color = "black", width = .5) +
  geom_point(aes(color = Subgroup), size = 2) + 
  facet_wrap(~Subgroup, scales = "free_x") +
  scale_x_discrete(labels = function(x) gsub("__.+$", "", x)) + 
  scale_y_continuous(limits = c(0,15)) +
  scale_color_brewer(palette = "RdBu") +
  theme_light() +
  theme(axis.text.x = element_text(angle = 45, vjust = .5),
        axis.title = element_text(face = "bold"),
        strip.background = element_rect(fill="grey70"),
        strip.text = element_text(color = 'white', face = "bold", size = "12"),
        legend.position = "none")  +
  geom_hline(yintercept = 0, size = 1) +
  labs(y = "Allele Substitution Effect (beta)")
```

```{r figure-5-7, out.width="100%", fig.cap='Top allele substitution effects (beta) by two breed subgroups.', fig.align='center'}
top_betas <- mlma %>% 
  filter(Subgroup_Code %in% c("cd", "cl", "cy", "dl", "dy", "ly")) %>% 
  filter(q < 0.10) %>%
  group_by(Plot_Code) %>%
  slice_max(order_by = b, n = 10) %>%
  arrange(-b)

top_betas %>% 
  mutate(SNP = reorder(SNP, b)) %>%
  group_by(Plot_Code, SNP) %>% 
  arrange(b) %>% 
  ungroup() %>% 
  mutate(SNP = factor(paste(SNP, Plot_Code, sep = "__"), 
                       levels = rev(paste(SNP, Plot_Code, sep = "__")))) %>%
ggplot(., aes(x = SNP, y = b)) +
  geom_errorbar(aes(ymin = b - se*1.96, ymax = b + se*1.96), color = "black", width = .5) +
  geom_point(aes(color = Plot_Code), size = 2) + 
  facet_wrap(~Plot_Code, scales = "free_x") +
  scale_x_discrete(labels = function(x) gsub("__.+$", "", x)) + 
  scale_y_continuous(limits = c(0,13)) +
  scale_color_brewer(palette = "RdBu") +
  theme_light() +
  theme(axis.text.x = element_text(angle = 45, vjust = .5),
        axis.title = element_text(face = "bold"),
        strip.background = element_rect(fill="grey70"),
        strip.text = element_text(color = 'white', face = "bold", size = "12"),
        legend.position = "none")  +
  geom_hline(yintercept = 0, size = 1) +
  labs(y = "Allele Substitution Effect (beta)")
```
```{r figure-5-8, out.width="100%", fig.cap='Top allele substitution effects (beta) by three+ breed subgroups.', fig.align='center'}
top_betas <- mlma %>% 
  filter(Subgroup_Code %in% c("dly", "cdly")) %>% 
  filter(q < 0.10) %>%
  group_by(Plot_Code) %>%
  slice_max(order_by = b, n = 10) %>%
  arrange(-b)

top_betas$Plot_Code <- factor(top_betas$Plot_Code,
                              levels = c("D, L, and Y", "C, D, L, and Y"))

top_betas %>% 
  mutate(SNP = reorder(SNP, b)) %>%
  group_by(Plot_Code, SNP) %>% 
  arrange(b) %>% 
  ungroup() %>% 
  mutate(SNP = factor(paste(SNP, Plot_Code, sep = "__"), 
                       levels = rev(paste(SNP, Plot_Code, sep = "__")))) %>%
ggplot(., aes(x = SNP, y = b)) +
  geom_errorbar(aes(ymin = b - se*1.96, ymax = b + se*1.96), color = "black", width = .5, size = .7) +
  geom_point(aes(color = Plot_Code), size = 2.5) + 
  facet_wrap(~Plot_Code, scales = "free_x") +
  scale_x_discrete(labels = function(x) gsub("__.+$", "", x)) + 
  scale_y_continuous(limits = c(0,5)) +
  scale_color_manual(values = c("firebrick3", "royalblue4")) +
  theme_light() +
  theme(axis.text.x = element_text(angle = 45, vjust = .5),
        axis.title = element_text(face = "bold"),
        strip.background = element_rect(fill="grey70"),
        strip.text = element_text(color = 'white', face = "bold", size = "12"),
        legend.position = "none")  +
  geom_hline(yintercept = 0, size = 1) +
  labs(y = "Allele Substitution Effect (beta)")
```

```{r figure-5-9, fig.align='center', fig.cap="Allele frequency changes over time for alleles with the largest allele substitution effect within each one breed subgroup.", out.width="90%"}
include_graphics("images/figure_5_9.PNG")
```

## QTL Enrichment

```{r figure-5-10, out.height="100%", fig.cap="Top 15 QTLs by enrichment P-value for the entire group of genotyped pigs.", fig.align='center'}
load("QTL_Enrich_Results.RData")
QTLenrich_plot(filter(qtl.enrich, Line == "All") %>%
                 top_n(15, -adj.pval), x = "QTL", pval = "adj.pval")
```

```{r figure-5-11, out.height="100%", fig.cap="Top 15 QTLs by enrichment P-value for the Duroc subgroup.", fig.align='center'}
load("QTL_Enrich_Results.RData")
QTLenrich_plot(filter(qtl.enrich, Line == "1006") %>%
                 top_n(15, -adj.pval), x = "QTL", pval = "adj.pval")
```

```{r figure-5-12, out.height="100%", fig.cap="Top 15 QTLs by enrichment P-value for the Landrace subgroup.", fig.align='center'}
load("QTL_Enrich_Results.RData")
QTLenrich_plot(filter(qtl.enrich, Line == "10") %>%
                 top_n(15, -adj.pval), x = "QTL", pval = "adj.pval")
```

```{r figure-5-13, out.height="100%", fig.cap="Top 15 QTLs by enrichment P-value for the Yorkshire subgroup.", fig.align='center'}
load("QTL_Enrich_Results.RData")
QTLenrich_plot(filter(qtl.enrich, Line == "11") %>%
                 top_n(15, -adj.pval), x = "QTL", pval = "adj.pval")
```

## Gene Enrichment

```{r figure-5-14, out.height="100%", fig.cap="Gene Ontology enrichment chart for the entire group of genotyped pigs.", fig.align='center'}
load("GeneAnnotation.RData")

gene_list <- gene.annotation %>%
  filter(Breed == "All") %>%
  select(gene_id)
gene_list <- gene_list$gene_id

gostres <- gost(query = gene_list, 
                organism = "sscrofa", ordered_query = FALSE, 
                multi_query = FALSE, significant = TRUE, exclude_iea = FALSE, 
                measure_underrepresentation = FALSE, evcodes = FALSE, 
                user_threshold = 0.05, correction_method = "fdr", 
                domain_scope = "annotated", custom_bg = NULL, 
                numeric_ns = "", sources = NULL, as_short_link = FALSE)

gostplot(gostres, capped = FALSE, interactive = FALSE)
```

```{r figure-5-15, out.height="100%", out.width="100%", fig.cap="Gene Ontology enrichment table for the entire group of genotyped pigs.", fig.align='center'}
publish_gosttable(gostres, highlight_terms = gostres$result,
                        use_colors = TRUE, 
                        show_columns = c("source", "term_name", "term_size", "intersection_size"),
                        filename = NULL)
```

## Principal Component Analysis

```{r input-pca}
pca <- fread("results/cdly/cdly_all_pca.eigenvec",
             select = c(1, 3:4),
             col.names = c("PigID", "PC1", "PC2")) %>%
  left_join(ped_plus_intmd[,c(1, 7)], by="PigID")

mendel_multigen <- fread("results/cdly/plink.mendel") %>%
  mutate(ERROR = paste(ERROR, V7, V8, V9, V10, sep = " ")) %>%
  select(FID:ERROR) %>%
  group_by(KID, CODE) %>%
  summarise(N = n()) %>%
  rename(PigID = KID,
         Code = CODE)

mendel_multigen_summary <- mendel_multigen %>%
  mutate(CodeCategory = ifelse(Code <= 2, "All",
                                ifelse(Code == 5 | Code == 8, "Pig",
                                       ifelse(Code == 3 | Code == 6, "Father & Pig", "Mother & Pig")))) %>%
  group_by(PigID, CodeCategory) %>%
  summarise(N = sum(N)) %>%
  mutate(ErrorRate = (N/46453) * 100,
         ErrorCategory = ifelse(ErrorRate >= 10, "High",
                                ifelse(ErrorRate < 5, "Low", "Moderate")),
         ErrorCategory = factor(ErrorCategory,
                                levels = c("Low", "Moderate", "High")))
```

```{r figure-5-16, out.height="100%", out.width="100%", fig.cap="Principal component analysis of all pigs", fig.align='center'}
ggplot(pca, aes(x = PC1, y = PC2, color = Line)) +
  geom_point() + 
  labs(x = "Principal Component 1",
       y = "Principal Component 2") +
  scale_color_npg() +
  theme_light() +
  theme(axis.title = element_text(face = "bold"),
        legend.title = element_text(face = "bold")) + 
  guides(color=guide_legend(title="Genetic Line"))
```

```{r figure-5-17, out.height="100%", out.width="100%", fig.cap="Principal component analysis of purebred pigs", fig.align='center'}
ggplot(pca %>% filter(Line %in% c("Duroc", "Landrace", "Yorkshire")), aes(x = PC1, y = PC2, color = Line)) +
  geom_point() + 
  labs(x = "Principal Component 1",
       y = "Principal Component 2") +
  scale_color_manual(values = c("Duroc" = "#4DBBD5FF",
                                "Landrace" = "#00A087FF",
                                "Yorkshire" = "#3C5488FF")) +
  theme_light() +
  theme(axis.title = element_text(face = "bold"),
        legend.title = element_text(face = "bold")) + 
  guides(color=guide_legend(title="Genetic Line"))
```

```{r figure-5-18, out.height="100%", out.width="100%", fig.cap="Mendelian errors", fig.align='center'}
mendel_multigen_summary %>%
  left_join(pca, by = "PigID") %>%
ggplot(., aes(x = PC1, y = PC2, color = ErrorCategory)) +
  geom_point() + 
  labs(x = "Principal Component 1",
       y = "Principal Component 2") +
  scale_color_manual(values = c("Low" = "#4DBBD5FF",
                                "Moderate" = "#3C5488FF",
                                "High" = "#E64B35FF")) +
  theme_light() +
  theme(axis.title = element_text(face = "bold"),
        legend.title = element_text(face = "bold")) +
  facet_wrap(~CodeCategory)
```
