# (APPENDIX) Appendix {-}

# Python Scripts

## CreateSubgroups.py {#createsubgroups}

**Script:**

```{python create-subgroups, eval=FALSE}
# /data/cjgwx7/tml_gpsm/scripts/python/CreateSubgroups.py

import pandas as pd

### Specify paths to necessary files for import
path_genotypes = "../../raw_data/level1/genotypes.txt"
path_ped_plus_intmd = "../../intmd/pedigree_map/ped_plus_intmd.txt"

### Specify paths for file export
path_s1 = "../../raw_data/level2/c.txt"
path_s2 = "../../raw_data/level2/d.txt"
path_s3 = "../../raw_data/level2/l.txt"
path_s4 = "../../raw_data/level2/y.txt"
path_s5 = "../../raw_data/level2/cd.txt"
path_s6 = "../../raw_data/level2/cl.txt"
path_s7 = "../../raw_data/level2/cy.txt"
path_s8 = "../../raw_data/level2/dl.txt"
path_s9 = "../../raw_data/level2/dy.txt"
path_s10 = "../../raw_data/level2/ly.txt"
path_s11 = "../../raw_data/level2/dly.txt"
path_s12 = "../../raw_data/level2/cdly.txt"
 
### Specify column names for input files
colnames_genotypes = ['PigID', 'SNPs']
colnames_ped_plus_intmd = ['PigID', 'Line']

### Read in genotype file and replace commas between SNP haplotypes
genotypes = pd.read_csv(path_genotypes, 
			sep = "\t", 
			names = colnames_genotypes, 
			header = None)

genotypes = genotypes.replace(',',
			      '', 
			      regex=True)

### Read in ped_plus_intmd
ped_plus_intmd = pd.read_csv(path_ped_plus_intmd, 
			     sep = " ", 
			     names = colnames_ped_plus_intmd, 
			     header = None, 
			     usecols = [0, 6])

print("Data Imported Successfully")

### Merge genotype and ped_plus_intmd to connect PigID to Line information
genotypes_plus = pd.merge(genotypes, 
	      		  ped_plus_intmd, 
	       		  how = 'left', 
	       		  on = 'PigID')
genotypes_plus = genotypes_plus[['PigID', 'Line']]
genotypes_plus = genotypes_plus.dropna()

### Create subgroups

## Crossbred
c = genotypes_plus[genotypes_plus.Line == 9006]

## Duroc
d = genotypes_plus[genotypes_plus.Line == 1006]

## Landrace
l = genotypes_plus[genotypes_plus.Line == 10]

## Yorkshire
y = genotypes_plus[genotypes_plus.Line == 11]

## Crossbred and Duroc
cd = genotypes_plus
cd['f'] = cd.Line.isin([9006, 1006])
cd = cd[cd.f == True]
cd = cd.drop(columns=['f'])

## Crossbred and Landrace
cl = genotypes_plus
cl['f'] = cl.Line.isin([9006, 10])
cl = cl[cl.f == True]
cl = cl.drop(columns=['f'])

## Crossbred and Yorkshire 
cy = genotypes_plus
cy['f'] = cy.Line.isin([9006, 11])
cy = cy[cy.f == True]
cy = cy.drop(columns=['f'])

## Duroc and Landrace
dl = genotypes_plus
dl['f'] = dl.Line.isin([1006, 10])
dl = dl[dl.f == True]
dl = dl.drop(columns=['f'])

## Duroc and Yorkshire
dy = genotypes_plus
dy['f'] = dy.Line.isin([1006, 11])
dy = dy[dy.f == True]
dy = dy.drop(columns=['f'])

## Landrace and Yorkshire
ly = genotypes_plus
ly['f'] = ly.Line.isin([10, 11])
ly = ly[ly.f == True]
ly = ly.drop(columns=['f'])

## Duroc, Landrace, and Yorkshire
dly = genotypes_plus
dly['f'] = dly.Line.isin([1006, 10, 11])
dly = dly[dly.f == True]
dly = dly.drop(columns=['f'])

## Crossbred, Duroc, Landrace, and Yorkshire
cdly = genotypes_plus
cdly['f'] = cdly.Line.isin([9006, 1006, 10, 11])
cdly = cdly[cdly.f == True]
cdly = cdly.drop(columns=['f'])

print("Subgroups Created Successfully")

### Merge subgroup IDs with genotypes
c = pd.merge(c['PigID'], genotypes, how='left', on='PigID')
d = pd.merge(d['PigID'], genotypes, how='left', on='PigID')
l = pd.merge(l['PigID'], genotypes, how='left', on='PigID')
y = pd.merge(y['PigID'], genotypes, how='left', on='PigID')
cd = pd.merge(cd['PigID'], genotypes, how='left', on='PigID')
cl = pd.merge(cl['PigID'], genotypes, how='left', on='PigID')
cy = pd.merge(cy['PigID'], genotypes, how='left', on='PigID')
dl = pd.merge(dl['PigID'], genotypes, how='left', on='PigID')
dy = pd.merge(dy['PigID'], genotypes, how='left', on='PigID')
ly = pd.merge(ly['PigID'], genotypes, how='left', on='PigID')
dly = pd.merge(dly['PigID'], genotypes, how='left', on='PigID')
cdly = pd.merge(cdly['PigID'], genotypes, how='left', on='PigID')

print("Genotype File Merged Successfully")

### Check file length (Number of pigs before quality control)
print("The length of Subgroup 1 is:", len(c))
print("The length of Subgroup 2 is:", len(d))
print("The length of Subgroup 3 is:", len(l))
print("The length of Subgroup 4 is:", len(y))
print("The length of Subgroup 5 is:", len(cd))
print("The length of Subgroup 6 is:", len(cl))
print("The length of Subgroup 7 is:", len(cy))
print("The length of Subgroup 8 is:", len(dl))
print("The length of Subgroup 9 is:", len(dy))
print("The length of Subgroup 10 is:", len(ly))
print("The length of Subgroup 11 is:", len(dly))
print("The length of Subgroup 12 is:", len(cdly))

print("Starting output of subgroup files")

c.to_csv(path_s1, sep = "\t", index = False, header = False)
d.to_csv(path_s2, sep = "\t", index = False, header = False)
l.to_csv(path_s3, sep = "\t", index = False, header = False)
y.to_csv(path_s4, sep = "\t", index = False, header = False)
cd.to_csv(path_s5, sep = "\t", index = False, header = False)
cl.to_csv(path_s6, sep = "\t", index = False, header = False)
cy.to_csv(path_s7, sep = "\t", index = False, header = False)
dl.to_csv(path_s8, sep = "\t", index = False, header = False)
dy.to_csv(path_s9, sep = "\t", index = False, header = False)
ly.to_csv(path_s10, sep = "\t", index = False, header = False)
dly.to_csv(path_s11, sep = "\t", index = False, header = False)
cdly.to_csv(path_s12, sep = "\t", index = False, header = False)

print("Genotype files for each subgroup successfully written")
```

**Log:**

```{}
Data Imported Successfully
Subgroups Created Successfully
Genotype File Merged Successfully
The length of Subgroup 1 is: 8532
The length of Subgroup 2 is: 16802
The length of Subgroup 3 is: 19342
The length of Subgroup 4 is: 18368
The length of Subgroup 5 is: 25334
The length of Subgroup 6 is: 27874
The length of Subgroup 7 is: 26900
The length of Subgroup 8 is: 36144
The length of Subgroup 9 is: 35170
The length of Subgroup 10 is: 37710
The length of Subgroup 11 is: 54512
The length of Subgroup 12 is: 63044
Starting output of subgroup files
Genotype files for each subgroup successfully written
```

# Bash Scripts

## batch.sh {#batch}

**Script:**

```{bash batch, eval=FALSE}
# /data/cjgwx7/tml_gpsm/scripts/bash/batch.sh

### Create fixed-width BLUPF90 genotype file for each subgroup
awk -F ' ' '{printf "%-15s%-50000s\n",$1,$2}' ../../raw_data/level2/c.txt > ../../raw_data/level2/c.txt_fw
awk -F ' ' '{printf "%-15s%-50000s\n",$1,$2}' ../../raw_data/level2/d.txt > ../../raw_data/level2/d.txt_fw
awk -F ' ' '{printf "%-15s%-50000s\n",$1,$2}' ../../raw_data/level2/l.txt > ../../raw_data/level2/l.txt_fw
awk -F ' ' '{printf "%-15s%-50000s\n",$1,$2}' ../../raw_data/level2/y.txt > ../../raw_data/level2/y.txt_fw
awk -F ' ' '{printf "%-15s%-50000s\n",$1,$2}' ../../raw_data/level2/cd.txt > ../../raw_data/level2/cd.txt_fw
awk -F ' ' '{printf "%-15s%-50000s\n",$1,$2}' ../../raw_data/level2/cl.txt > ../../raw_data/level2/cl.txt_fw
awk -F ' ' '{printf "%-15s%-50000s\n",$1,$2}' ../../raw_data/level2/cy.txt > ../../raw_data/level2/cy.txt_fw
awk -F ' ' '{printf "%-15s%-50000s\n",$1,$2}' ../../raw_data/level2/dl.txt > ../../raw_data/level2/dl.txt_fw
awk -F ' ' '{printf "%-15s%-50000s\n",$1,$2}' ../../raw_data/level2/dy.txt > ../../raw_data/level2/dy.txt_fw
awk -F ' ' '{printf "%-15s%-50000s\n",$1,$2}' ../../raw_data/level2/ly.txt > ../../raw_data/level2/ly.txt_fw
awk -F ' ' '{printf "%-15s%-50000s\n",$1,$2}' ../../raw_data/level2/dly.txt > ../../raw_data/level2/dly.txt_fw
awk -F ' ' '{printf "%-15s%-50000s\n",$1,$2}' ../../raw_data/level2/cdly.txt > ../../raw_data/level2/cdly.txt_fw

### Delete old .txt files
rm ../../raw_data/level2/*.txt

### Split files into batches of < 25,000 observations each
### Subgroup 1
cp ../../raw_data/level2/c.txt_fw ../../raw_data/level3/c/c.aa

### Subgroup 2
cp ../../raw_data/level2/d.txt_fw ../../raw_data/level3/d/d.aa

### Subgroup 3
cp ../../raw_data/level2/l.txt_fw ../../raw_data/level3/l/l.aa

### Subgroup 4
cp ../../raw_data/level2/y.txt_fw ../../raw_data/level3/y/y.aa

### Subgroup 5
split --verbose ../../raw_data/level2/cd.txt_fw -l12667 cd.
mv cd* ../../raw_data/level3/cd/

### Subgroup 6
split --verbose ../../raw_data/level2/cl.txt_fw -l13937 cl.
mv cl* ../../raw_data/level3/cl/

### Subgroup 7
split --verbose ../../raw_data/level2/cy.txt_fw -l13450 cy.
mv cy* ../../raw_data/level3/cy/

### Subgroup 8
split --verbose ../../raw_data/level2/dl.txt_fw -l18072 dl.
mv dl* ../../raw_data/level3/dl/

### Subgroup 9
split --verbose ../../raw_data/level2/dy.txt_fw -l17585 dy.
mv dy* ../../raw_data/level3/dy/

### Subgroup 10
split --verbose ../../raw_data/level2/ly.txt_fw -l18855 ly.
mv ly* ../../raw_data/level3/ly/

### Subgroup 11
split --verbose ../../raw_data/level2/dly.txt_fw -l18171 dly.
mv dly* ../../raw_data/level3/dly/

### Subgroup 12
split --verbose ../../raw_data/level2/cdly.txt_fw -l21015 cdly.
mv cdly* ../../raw_data/level3/cdly/
```

**Log:**

```{}
creating file `cd.aa'
creating file `cd.ab'
creating file `cl.aa'
creating file `cl.ab'
creating file `cy.aa'
creating file `cy.ab'
creating file `dl.aa'
creating file `dl.ab'
creating file `dy.aa'
creating file `dy.ab'
creating file `ly.aa'
creating file `ly.ab'
creating file `dly.aa'
creating file `dly.ab'
creating file `dly.ac'
creating file `cdly.aa'
creating file `cdly.ab'
creating file `cdly.ac'
```

## renum.sh {#renum}

**Script:**

```{bash renum, eval=FALSE}
# /data/cjgwx7/tml_gpsm/scripts/bash/renum.sh

### Copy necessary files into current directory
cp ../../intmd/pedigree_map/{ped_plus_intmd.txt,ped_intmd.txt,map_intmd.txt} .

### Copy genotype file into current directory
cp ../../raw_data/level3/c/c.aa .

### Run renumF90
renumf90 ../blupf90/renum.c_aa

### Save PLINK PED and MAP files
preGSf90 renf90.par

### Move important files to new location
mv renf90.dat toPlink.map toPlink.ped renadd01.ped c.aa_XrefID sum2pq freqdata.count Gen_call_rate ../../raw_data/level4/c/

### Remove unnecessary files
rm c.aa renf90.tables renf90.fields renf90.par

### Repeat above for each subgroup/batch combination

### Subgroup 2
cp ../../raw_data/level3/d/d.aa .
renumf90 ../blupf90/renum.d_aa
preGSf90 renf90.par
mv renf90.dat toPlink.map toPlink.ped renadd01.ped d.aa_XrefID sum2pq freqdata.count Gen_call_rate ../../raw_data/level4/d/
rm d.aa renf90.tables renf90.fields renf90.par

### Subgroup 3
cp ../../raw_data/level3/l/l.aa .
renumf90 ../blupf90/renum.l_aa
preGSf90 renf90.par
mv renf90.dat toPlink.map toPlink.ped renadd01.ped l.aa_XrefID sum2pq freqdata.count Gen_call_rate ../../raw_data/level4/l/
rm l.aa renf90.tables renf90.fields renf90.par

### Subgroup 4
cp ../../raw_data/level3/y/y.aa .
renumf90 ../blupf90/renum.y_aa
preGSf90 renf90.par
mv renf90.dat toPlink.map toPlink.ped renadd01.ped y.aa_XrefID sum2pq freqdata.count Gen_call_rate ../../raw_data/level4/y/
rm y.aa renf90.tables renf90.fields renf90.par

### Subgroup 5
cp ../../raw_data/level3/cd/cd.aa .
renumf90 ../blupf90/renum.cd_aa
preGSf90 renf90.par
mv renf90.dat toPlink.map toPlink.ped renadd01.ped cd.aa_XrefID sum2pq freqdata.count Gen_call_rate ../../raw_data/level4/cd/aa/
rm cd.aa renf90.tables renf90.fields renf90.par

cp ../../raw_data/level3/cd/cd.ab .
renumf90 ../blupf90/renum.cd_ab
preGSf90 renf90.par
mv renf90.dat toPlink.map toPlink.ped renadd01.ped cd.ab_XrefID sum2pq freqdata.count Gen_call_rate ../../raw_data/level4/cd/ab/
rm cd.ab renf90.tables renf90.fields renf90.par

### Subgroup 6
cp ../../raw_data/level3/cl/cl.aa .
renumf90 ../blupf90/renum.cl_aa
preGSf90 renf90.par
mv renf90.dat toPlink.map toPlink.ped renadd01.ped cl.aa_XrefID sum2pq freqdata.count Gen_call_rate ../../raw_data/level4/cl/aa/
rm cl.aa renf90.tables renf90.fields renf90.par

cp ../../raw_data/level3/cl/cl.ab .
renumf90 ../blupf90/renum.cl_ab
preGSf90 renf90.par
mv renf90.dat toPlink.map toPlink.ped renadd01.ped cl.ab_XrefID sum2pq freqdata.count Gen_call_rate ../../raw_data/level4/cl/ab/
rm cl.ab renf90.tables renf90.fields renf90.par

### Subgroup 7
cp ../../raw_data/level3/cy/cy.aa .
renumf90 ../blupf90/renum.cy_aa
preGSf90 renf90.par
mv renf90.dat toPlink.map toPlink.ped renadd01.ped cy.aa_XrefID sum2pq freqdata.count Gen_call_rate ../../raw_data/level4/cy/aa/
rm cy.aa renf90.tables renf90.fields renf90.par

cp ../../raw_data/level3/cy/cy.ab .
renumf90 ../blupf90/renum.cy_ab
preGSf90 renf90.par
mv renf90.dat toPlink.map toPlink.ped renadd01.ped cy.ab_XrefID sum2pq freqdata.count Gen_call_rate ../../raw_data/level4/cy/ab/
rm cy.ab renf90.tables renf90.fields renf90.par

### Subgroup 8
cp ../../raw_data/level3/dl/dl.aa .
renumf90 ../blupf90/renum.dl_aa
preGSf90 renf90.par
mv renf90.dat toPlink.map toPlink.ped renadd01.ped dl.aa_XrefID sum2pq freqdata.count Gen_call_rate ../../raw_data/level4/dl/aa/
rm dl.aa renf90.tables renf90.fields renf90.par

cp ../../raw_data/level3/dl/dl.ab .
renumf90 ../blupf90/renum.dl_ab
preGSf90 renf90.par
mv renf90.dat toPlink.map toPlink.ped renadd01.ped dl.ab_XrefID sum2pq freqdata.count Gen_call_rate ../../raw_data/level4/dl/ab/
rm dl.ab renf90.tables renf90.fields renf90.par

### Subgroup 9
cp ../../raw_data/level3/dy/dy.aa .
renumf90 ../blupf90/renum.dy_aa
preGSf90 renf90.par
mv renf90.dat toPlink.map toPlink.ped renadd01.ped dy.aa_XrefID sum2pq freqdata.count Gen_call_rate ../../raw_data/level4/dy/aa/
rm dy.aa renf90.tables renf90.fields renf90.par

cp ../../raw_data/level3/dy/dy.ab .
renumf90 ../blupf90/renum.dy_ab
preGSf90 renf90.par
mv renf90.dat toPlink.map toPlink.ped renadd01.ped dy.ab_XrefID sum2pq freqdata.count Gen_call_rate ../../raw_data/level4/dy/ab/
rm dy.ab renf90.tables renf90.fields renf90.par

### Subgroup 10
cp ../../raw_data/level3/ly/ly.aa .
renumf90 ../blupf90/renum.ly_aa
preGSf90 renf90.par
mv renf90.dat toPlink.map toPlink.ped renadd01.ped ly.aa_XrefID sum2pq freqdata.count Gen_call_rate ../../raw_data/level4/ly/aa/
rm ly.aa renf90.tables renf90.fields renf90.par

cp ../../raw_data/level3/ly/ly.ab .
renumf90 ../blupf90/renum.ly_ab
preGSf90 renf90.par
mv renf90.dat toPlink.map toPlink.ped renadd01.ped ly.ab_XrefID sum2pq freqdata.count Gen_call_rate ../../raw_data/level4/ly/ab/
rm ly.ab renf90.tables renf90.fields renf90.par

### Subgroup 11
cp ../../raw_data/level3/dly/dly.aa .
renumf90 ../blupf90/renum.dly_aa
preGSf90 renf90.par
mv renf90.dat toPlink.map toPlink.ped renadd01.ped dly.aa_XrefID sum2pq freqdata.count Gen_call_rate ../../raw_data/level4/dly/aa/
rm dly.aa renf90.tables renf90.fields renf90.par

cp ../../raw_data/level3/dly/dly.ab .
renumf90 ../blupf90/renum.dly_ab
preGSf90 renf90.par
mv renf90.dat toPlink.map toPlink.ped renadd01.ped dly.ab_XrefID sum2pq freqdata.count Gen_call_rate ../../raw_data/level4/dly/ab/
rm dly.ab renf90.tables renf90.fields renf90.par

cp ../../raw_data/level3/dly/dly.ac .
renumf90 ../blupf90/renum.dly_ac
preGSf90 renf90.par
mv renf90.dat toPlink.map toPlink.ped renadd01.ped dly.ac_XrefID sum2pq freqdata.count Gen_call_rate ../../raw_data/level4/dly/ac/
rm dly.ac renf90.tables renf90.fields renf90.par

### Subgroup 12
cp ../../raw_data/level3/cdly/cdly.aa .
renumf90 ../blupf90/renum.cdly_aa
preGSf90 renf90.par
mv renf90.dat toPlink.map toPlink.ped renadd01.ped cdly.aa_XrefID sum2pq freqdata.count Gen_call_rate ../../raw_data/level4/cdly/aa/
rm cdly.aa renf90.tables renf90.fields renf90.par

cp ../../raw_data/level3/cdly/cdly.ab .
renumf90 ../blupf90/renum.cdly_ab
preGSf90 renf90.par
mv renf90.dat toPlink.map toPlink.ped renadd01.ped cdly.ab_XrefID sum2pq freqdata.count Gen_call_rate ../../raw_data/level4/cdly/ab/
rm cdly.ab renf90.tables renf90.fields renf90.par

cp ../../raw_data/level3/cdly/cdly.ac .
renumf90 ../blupf90/renum.cdly_ac
preGSf90 renf90.par
mv renf90.dat toPlink.map toPlink.ped renadd01.ped cdly.ac_XrefID sum2pq freqdata.count Gen_call_rate ../../raw_data/level4/cdly/ac/
rm cdly.ac renf90.tables renf90.fields renf90.par

### Remove unnecessary files
rm ped_plus_intmd.txt ped_intmd.txt map_intmd.txt
```

## ProcessPlinkF90.sh {#processplinkf90}

**Script:**

```{bash processplinkf90, eval=FALSE}
# /data/cjgwx7/tml_gpsm/scripts/bash/ProcessPlinkF90.sh

### Subgroup 1
### Remove extra blank column
cut -d ' ' -f 1-6,8- ../../raw_data/level4/c/toPlink.ped > ../../raw_data/level4/c/toPlink_v2.ped

### Extract genotypes
cut -d ' ' -f 7- ../../raw_data/level4/c/toPlink_v2.ped > ../../raw_data/level4/c/toPlink_geno.ped

### Replace renumbered IDs with actual IDs
Rscript ../r/ProcessPlinkF90/c_aa.R

### Column bind columns 1-6 and the genotypes
paste -d ' ' ../../raw_data/level4/c/c_aa.fam ../../raw_data/level4/c/toPlink_geno.ped > ../../raw_data/level6/c.ped

### Subgroup 2
cut -d ' ' -f 1-6,8- ../../raw_data/level4/d/toPlink.ped > ../../raw_data/level4/d/toPlink_v2.ped
cut -d ' ' -f 7- ../../raw_data/level4/d/toPlink_v2.ped > ../../raw_data/level4/d/toPlink_geno.ped

Rscript ../r/ProcessPlinkF90/d_aa.R

paste -d ' ' ../../raw_data/level4/d/d_aa.fam ../../raw_data/level4/d/toPlink_geno.ped > ../../raw_data/level6/d.ped

### Subgroup 3
cut -d ' ' -f 1-6,8- ../../raw_data/level4/l/toPlink.ped > ../../raw_data/level4/l/toPlink_v2.ped
cut -d ' ' -f 7- ../../raw_data/level4/l/toPlink_v2.ped > ../../raw_data/level4/l/toPlink_geno.ped

Rscript ../r/ProcessPlinkF90/l_aa.R

paste -d ' ' ../../raw_data/level4/l/l_aa.fam ../../raw_data/level4/l/toPlink_geno.ped > ../../raw_data/level6/l.ped

### Subgroup 4
cut -d ' ' -f 1-6,8- ../../raw_data/level4/y/toPlink.ped > ../../raw_data/level4/y/toPlink_v2.ped
cut -d ' ' -f 7- ../../raw_data/level4/y/toPlink_v2.ped > ../../raw_data/level4/y/toPlink_geno.ped

Rscript ../r/ProcessPlinkF90/y_aa.R

paste -d ' ' ../../raw_data/level4/y/y_aa.fam ../../raw_data/level4/y/toPlink_geno.ped > ../../raw_data/level6/y.ped

### Subgroup 5
### Batch 1
cut -d ' ' -f 1-6,8- ../../raw_data/level4/cd/aa/toPlink.ped > ../../raw_data/level4/cd/aa/toPlink_v2.ped
cut -d ' ' -f 7- ../../raw_data/level4/cd/aa/toPlink_v2.ped > ../../raw_data/level4/cd/aa/toPlink_geno.ped

Rscript ../r/ProcessPlinkF90/cd_aa.R

paste -d ' ' ../../raw_data/level4/cd/aa/cd_aa.fam ../../raw_data/level4/cd/aa/toPlink_geno.ped > ../../raw_data/level5/cd/cd_aa.ped

### Batch 2
cut -d ' ' -f 1-6,8- ../../raw_data/level4/cd/ab/toPlink.ped > ../../raw_data/level4/cd/ab/toPlink_v2.ped
cut -d ' ' -f 7- ../../raw_data/level4/cd/ab/toPlink_v2.ped > ../../raw_data/level4/cd/ab/toPlink_geno.ped

Rscript ../r/ProcessPlinkF90/cd_ab.R

paste -d ' ' ../../raw_data/level4/cd/ab/cd_ab.fam ../../raw_data/level4/cd/ab/toPlink_geno.ped > ../../raw_data/level5/cd/cd_ab.ped

cat ../../raw_data/level5/cd/cd_aa.ped ../../raw_data/level5/cd/cd_ab.ped > ../../raw_data/level6/cd.ped

### Subgroup 6
### Batch 1
cut -d ' ' -f 1-6,8- ../../raw_data/level4/cl/aa/toPlink.ped > ../../raw_data/level4/cl/aa/toPlink_v2.ped
cut -d ' ' -f 7- ../../raw_data/level4/cl/aa/toPlink_v2.ped > ../../raw_data/level4/cl/aa/toPlink_geno.ped

Rscript ../r/ProcessPlinkF90/cl_aa.R

paste -d ' ' ../../raw_data/level4/cl/aa/cl_aa.fam ../../raw_data/level4/cl/aa/toPlink_geno.ped > ../../raw_data/level5/cl/cl_aa.ped

### Batch 2
cut -d ' ' -f 1-6,8- ../../raw_data/level4/cl/ab/toPlink.ped > ../../raw_data/level4/cl/ab/toPlink_v2.ped
cut -d ' ' -f 7- ../../raw_data/level4/cl/ab/toPlink_v2.ped > ../../raw_data/level4/cl/ab/toPlink_geno.ped

Rscript ../r/ProcessPlinkF90/cl_ab.R

paste -d ' ' ../../raw_data/level4/cl/ab/cl_ab.fam ../../raw_data/level4/cl/ab/toPlink_geno.ped > ../../raw_data/level5/cl/cl_ab.ped

cat ../../raw_data/level5/cl/cl_aa.ped ../../raw_data/level5/cl/cl_ab.ped > ../../raw_data/level6/cl.ped

### Subgroup 7
### Batch 1
cut -d ' ' -f 1-6,8- ../../raw_data/level4/cy/aa/toPlink.ped > ../../raw_data/level4/cy/aa/toPlink_v2.ped
cut -d ' ' -f 7- ../../raw_data/level4/cy/aa/toPlink_v2.ped > ../../raw_data/level4/cy/aa/toPlink_geno.ped

Rscript ../r/ProcessPlinkF90/cy_aa.R

paste -d ' ' ../../raw_data/level4/cy/aa/cy_aa.fam ../../raw_data/level4/cy/aa/toPlink_geno.ped > ../../raw_data/level5/cy/cy_aa.ped

### Batch 2
cut -d ' ' -f 1-6,8- ../../raw_data/level4/cy/ab/toPlink.ped > ../../raw_data/level4/cy/ab/toPlink_v2.ped
cut -d ' ' -f 7- ../../raw_data/level4/cy/ab/toPlink_v2.ped > ../../raw_data/level4/cy/ab/toPlink_geno.ped

Rscript ../r/ProcessPlinkF90/cy_ab.R

paste -d ' ' ../../raw_data/level4/cy/ab/cy_ab.fam ../../raw_data/level4/cy/ab/toPlink_geno.ped > ../../raw_data/level5/cy/cy_ab.ped

cat ../../raw_data/level5/cy/cy_aa.ped ../../raw_data/level5/cy/cy_ab.ped > ../../raw_data/level6/cy.ped

### Subgroup 8
### Batch 1
cut -d ' ' -f 1-6,8- ../../raw_data/level4/dl/aa/toPlink.ped > ../../raw_data/level4/dl/aa/toPlink_v2.ped
cut -d ' ' -f 7- ../../raw_data/level4/dl/aa/toPlink_v2.ped > ../../raw_data/level4/dl/aa/toPlink_geno.ped

Rscript ../r/ProcessPlinkF90/dl_aa.R

paste -d ' ' ../../raw_data/level4/dl/aa/dl_aa.fam ../../raw_data/level4/dl/aa/toPlink_geno.ped > ../../raw_data/level5/dl/dl_aa.ped

### Batch 2
cut -d ' ' -f 1-6,8- ../../raw_data/level4/dl/ab/toPlink.ped > ../../raw_data/level4/dl/ab/toPlink_v2.ped
cut -d ' ' -f 7- ../../raw_data/level4/dl/ab/toPlink_v2.ped > ../../raw_data/level4/dl/ab/toPlink_geno.ped

Rscript ../r/ProcessPlinkF90/dl_ab.R

paste -d ' ' ../../raw_data/level4/dl/ab/dl_ab.fam ../../raw_data/level4/dl/ab/toPlink_geno.ped > ../../raw_data/level5/dl/dl_ab.ped

cat ../../raw_data/level5/dl/dl_aa.ped ../../raw_data/level5/dl/dl_ab.ped > ../../raw_data/level6/dl.ped

### Subgroup 9
### Batch 1
cut -d ' ' -f 1-6,8- ../../raw_data/level4/dy/aa/toPlink.ped > ../../raw_data/level4/dy/aa/toPlink_v2.ped
cut -d ' ' -f 7- ../../raw_data/level4/dy/aa/toPlink_v2.ped > ../../raw_data/level4/dy/aa/toPlink_geno.ped

Rscript ../r/ProcessPlinkF90/dy_aa.R

paste -d ' ' ../../raw_data/level4/dy/aa/dy_aa.fam ../../raw_data/level4/dy/aa/toPlink_geno.ped > ../../raw_data/level5/dy/dy_aa.ped

### Batch 2
cut -d ' ' -f 1-6,8- ../../raw_data/level4/dy/ab/toPlink.ped > ../../raw_data/level4/dy/ab/toPlink_v2.ped
cut -d ' ' -f 7- ../../raw_data/level4/dy/ab/toPlink_v2.ped > ../../raw_data/level4/dy/ab/toPlink_geno.ped

Rscript ../r/ProcessPlinkF90/dy_ab.R

paste -d ' ' ../../raw_data/level4/dy/ab/dy_ab.fam ../../raw_data/level4/dy/ab/toPlink_geno.ped > ../../raw_data/level5/dy/dy_ab.ped

cat ../../raw_data/level5/dy/dy_aa.ped ../../raw_data/level5/dy/dy_ab.ped > ../../raw_data/level6/dy.ped

### Subgroup 10
### Batch 1
cut -d ' ' -f 1-6,8- ../../raw_data/level4/ly/aa/toPlink.ped > ../../raw_data/level4/ly/aa/toPlink_v2.ped
cut -d ' ' -f 7- ../../raw_data/level4/ly/aa/toPlink_v2.ped > ../../raw_data/level4/ly/aa/toPlink_geno.ped

Rscript ../r/ProcessPlinkF90/ly_aa.R

paste -d ' ' ../../raw_data/level4/ly/aa/ly_aa.fam ../../raw_data/level4/ly/aa/toPlink_geno.ped > ../../raw_data/level5/ly/ly_aa.ped

### Batch 2
cut -d ' ' -f 1-6,8- ../../raw_data/level4/ly/ab/toPlink.ped > ../../raw_data/level4/ly/ab/toPlink_v2.ped
cut -d ' ' -f 7- ../../raw_data/level4/ly/ab/toPlink_v2.ped > ../../raw_data/level4/ly/ab/toPlink_geno.ped

Rscript ../r/ProcessPlinkF90/ly_ab.R

paste -d ' ' ../../raw_data/level4/ly/ab/ly_ab.fam ../../raw_data/level4/ly/ab/toPlink_geno.ped > ../../raw_data/level5/ly/ly_ab.ped

cat ../../raw_data/level5/ly/ly_aa.ped ../../raw_data/level5/ly/ly_ab.ped > ../../raw_data/level6/ly.ped

### Subgroup 11
### Batch 1
cut -d ' ' -f 1-6,8- ../../raw_data/level4/dly/aa/toPlink.ped > ../../raw_data/level4/dly/aa/toPlink_v2.ped
cut -d ' ' -f 7- ../../raw_data/level4/dly/aa/toPlink_v2.ped > ../../raw_data/level4/dly/aa/toPlink_geno.ped

Rscript ../r/ProcessPlinkF90/dly_aa.R

paste -d ' ' ../../raw_data/level4/dly/aa/dly_aa.fam ../../raw_data/level4/dly/aa/toPlink_geno.ped > ../../raw_data/level5/dly/dly_aa.ped

### Batch 2
cut -d ' ' -f 1-6,8- ../../raw_data/level4/dly/ab/toPlink.ped > ../../raw_data/level4/dly/ab/toPlink_v2.ped
cut -d ' ' -f 7- ../../raw_data/level4/dly/ab/toPlink_v2.ped > ../../raw_data/level4/dly/ab/toPlink_geno.ped

Rscript ../r/ProcessPlinkF90/dly_ab.R

paste -d ' ' ../../raw_data/level4/dly/ab/dly_ab.fam ../../raw_data/level4/dly/ab/toPlink_geno.ped > ../../raw_data/level5/dly/dly_ab.ped

### Batch 3
cut -d ' ' -f 1-6,8- ../../raw_data/level4/dly/ac/toPlink.ped > ../../raw_data/level4/dly/ac/toPlink_v2.ped
cut -d ' ' -f 7- ../../raw_data/level4/dly/ac/toPlink_v2.ped > ../../raw_data/level4/dly/ac/toPlink_geno.ped

Rscript ../r/ProcessPlinkF90/dly_ac.R

paste -d ' ' ../../raw_data/level4/dly/ac/dly_ac.fam ../../raw_data/level4/dly/ac/toPlink_geno.ped > ../../raw_data/level5/dly/dly_ac.ped

cat ../../raw_data/level5/dly/dly_aa.ped ../../raw_data/level5/dly/dly_ab.ped ../../raw_data/level5/dly/dly_ac.ped > ../../raw_data/level6/dly.ped

### Subgroup 12
### Batch 1
cut -d ' ' -f 1-6,8- ../../raw_data/level4/cdly/aa/toPlink.ped > ../../raw_data/level4/cdly/aa/toPlink_v2.ped
cut -d ' ' -f 7- ../../raw_data/level4/cdly/aa/toPlink_v2.ped > ../../raw_data/level4/cdly/aa/toPlink_geno.ped

Rscript ../r/ProcessPlinkF90/cdly_aa.R

paste -d ' ' ../../raw_data/level4/cdly/aa/cdly_aa.fam ../../raw_data/level4/cdly/aa/toPlink_geno.ped > ../../raw_data/level5/cdly/cdly_aa.ped

### Batch 2
cut -d ' ' -f 1-6,8- ../../raw_data/level4/cdly/ab/toPlink.ped > ../../raw_data/level4/cdly/ab/toPlink_v2.ped
cut -d ' ' -f 7- ../../raw_data/level4/cdly/ab/toPlink_v2.ped > ../../raw_data/level4/cdly/ab/toPlink_geno.ped

Rscript ../r/ProcessPlinkF90/cdly_ab.R

paste -d ' ' ../../raw_data/level4/cdly/ab/cdly_ab.fam ../../raw_data/level4/cdly/ab/toPlink_geno.ped > ../../raw_data/level5/cdly/cdly_ab.ped

### Batch 3
cut -d ' ' -f 1-6,8- ../../raw_data/level4/cdly/ac/toPlink.ped > ../../raw_data/level4/cdly/ac/toPlink_v2.ped
cut -d ' ' -f 7- ../../raw_data/level4/cdly/ac/toPlink_v2.ped > ../../raw_data/level4/cdly/ac/toPlink_geno.ped

Rscript ../r/ProcessPlinkF90/cdly_ac.R

paste -d ' ' ../../raw_data/level4/cdly/ac/cdly_ac.fam ../../raw_data/level4/cdly/ac/toPlink_geno.ped > ../../raw_data/level5/cdly/cdly_ac.ped

cat ../../raw_data/level5/cdly/cdly_aa.ped ../../raw_data/level5/cdly/cdly_ab.ped ../../raw_data/level5/cdly/cdly_ac.ped > ../../raw_data/level6/cdly.ped
```

## MakeBed.sh {#make-bed}

**Script:**

```{bash make-bed, eval=FALSE}
# /data/tml_gpsm/cjgwx7/scripts/bash/

### Generate BED, BIM, and FAM files using PLINK software
### Exclude sex chromosomes (X and Y/19 and 20)
plink --ped ../../raw_data/level6/c.ped --map ../../raw_data/level6/global.map --out ../../raw_data/level7/c/c --make-bed --not-chr 19-20
plink --ped ../../raw_data/level6/d.ped --map ../../raw_data/level6/global.map --out ../../raw_data/level7/d/d --make-bed --not-chr 19-20
plink --ped ../../raw_data/level6/l.ped --map ../../raw_data/level6/global.map --out ../../raw_data/level7/l/l --make-bed --not-chr 19-20
plink --ped ../../raw_data/level6/y.ped --map ../../raw_data/level6/global.map --out ../../raw_data/level7/y/y --make-bed --not-chr 19-20
plink --ped ../../raw_data/level6/cd.ped --map ../../raw_data/level6/global.map --out ../../raw_data/level7/cd/cd --make-bed --not-chr 19-20
plink --ped ../../raw_data/level6/cl.ped --map ../../raw_data/level6/global.map --out ../../raw_data/level7/cl/cl --make-bed --not-chr 19-20
plink --ped ../../raw_data/level6/cy.ped --map ../../raw_data/level6/global.map --out ../../raw_data/level7/cy/cy --make-bed --not-chr 19-20
plink --ped ../../raw_data/level6/dl.ped --map ../../raw_data/level6/global.map --out ../../raw_data/level7/dl/dl --make-bed --not-chr 19-20
plink --ped ../../raw_data/level6/dy.ped --map ../../raw_data/level6/global.map --out ../../raw_data/level7/dy/dy --make-bed --not-chr 19-20
plink --ped ../../raw_data/level6/ly.ped --map ../../raw_data/level6/global.map --out ../../raw_data/level7/ly/ly --make-bed --not-chr 19-20
plink --ped ../../raw_data/level6/dly.ped --map ../../raw_data/level6/global.map --out ../../raw_data/level7/dly/dly --make-bed --not-chr 19-20
plink --ped ../../raw_data/level6/cdly.ped --map ../../raw_data/level6/global.map --out ../../raw_data/level7/cdly/cdly --make-bed --not-chr 19-20
```

## QualityControl.sh {#quality-control-2}

**Script:**

```{bash quality-control, eval=FALSE}
# /data/cjgwx7/tml_gpsm/scripts/bash/

### Filter by SNPs
### Subgroup 1
plink --bfile ../../raw_data/level7/c/c --out ../../intmd/input_data/c/c --make-bed --maf 0.01 --geno 0.1 --nonfounders

### Filter by pigs
plink --bfile ../../intmd/input_data/c/c --out ../../intmd/input_data/c/c --make-bed --mind 0.1 --nonfounders

### Subgroup 2
plink --bfile ../../raw_data/level7/d/d --out ../../intmd/input_data/d/d --make-bed --maf 0.01 --geno 0.1 --nonfounders
plink --bfile ../../intmd/input_data/d/d --out ../../intmd/input_data/d/d --make-bed --mind 0.1 --nonfounders

### Subgroup 3
plink --bfile ../../raw_data/level7/l/l --out ../../intmd/input_data/l/l --make-bed --maf 0.01 --geno 0.1 --nonfounders
plink --bfile ../../intmd/input_data/l/l --out ../../intmd/input_data/l/l --make-bed --mind 0.1 --nonfounders

### Subgroup 4
plink --bfile ../../raw_data/level7/y/y --out ../../intmd/input_data/y/y --make-bed --maf 0.01 --geno 0.1 --nonfounders
plink --bfile ../../intmd/input_data/y/y --out ../../intmd/input_data/y/y --make-bed --mind 0.1 --nonfounders

### Subgroup 5
plink --bfile ../../raw_data/level7/cd/cd --out ../../intmd/input_data/cd/cd --make-bed --maf 0.01 --geno 0.1 --nonfounders
plink --bfile ../../intmd/input_data/cd/cd --out ../../intmd/input_data/cd/cd --make-bed --mind 0.1 --nonfounders

### Subgroup 6
plink --bfile ../../raw_data/level7/cl/cl --out ../../intmd/input_data/cl/cl --make-bed --maf 0.01 --geno 0.1 --nonfounders
plink --bfile ../../intmd/input_data/cl/cl --out ../../intmd/input_data/cl/cl --make-bed --mind 0.1 --nonfounders

### Subgroup 7
plink --bfile ../../raw_data/level7/cy/cy --out ../../intmd/input_data/cy/cy --make-bed --maf 0.01 --geno 0.1 --nonfounders
plink --bfile ../../intmd/input_data/cy/cy --out ../../intmd/input_data/cy/cy --make-bed --mind 0.1 --nonfounders

### Subgroup 8
plink --bfile ../../raw_data/level7/dl/dl --out ../../intmd/input_data/dl/dl --make-bed --maf 0.01 --geno 0.1 --nonfounders
plink --bfile ../../intmd/input_data/dl/dl --out ../../intmd/input_data/dl/dl --make-bed --mind 0.1 --nonfounders

### Subgroup 9
plink --bfile ../../raw_data/level7/dy/dy --out ../../intmd/input_data/dy/dy --make-bed --maf 0.01 --geno 0.1 --nonfounders
plink --bfile ../../intmd/input_data/dy/dy --out ../../intmd/input_data/dy/dy --make-bed --mind 0.1 --nonfounders

### Subgroup 10
plink --bfile ../../raw_data/level7/ly/ly --out ../../intmd/input_data/ly/ly --make-bed --maf 0.01 --geno 0.1 --nonfounders
plink --bfile ../../intmd/input_data/ly/ly --out ../../intmd/input_data/ly/ly --make-bed --mind 0.1 --nonfounders

### Subgroup 11
plink --bfile ../../raw_data/level7/dly/dly --out ../../intmd/input_data/dly/dly --make-bed --maf 0.01 --geno 0.1 --nonfounders
plink --bfile ../../intmd/input_data/dly/dly --out ../../intmd/input_data/dly/dly --make-bed --mind 0.1 --nonfounders

### Subgroup 12
plink --bfile ../../raw_data/level7/cdly/cdly --out ../../intmd/input_data/cdly/cdly --make-bed --maf 0.01 --geno 0.1 --nonfounders
plink --bfile ../../intmd/input_data/cdly/cdly --out ../../intmd/input_data/cdly/cdly --make-bed --mind 0.1 --nonfounders
```

## MakeGRMs.sh {#makegrms}

**Script:**

```{bash makegrms, eval=FALSE}
# /data/cjgwx7/tml_gpsm/scripts/bash/

### Make Genomic Relationship Matrix for each subset

gcta64 --bfile ../../intmd/input_data/c/c --make-grm --out ../../intmd/grms/c/c_all --thread-num 24
gcta64 --bfile ../../intmd/input_data/d/d --make-grm --out ../../intmd/grms/d/d_all --thread-num 24
gcta64 --bfile ../../intmd/input_data/l/l --make-grm --out ../../intmd/grms/l/l_all --thread-num 24
gcta64 --bfile ../../intmd/input_data/y/y --make-grm --out ../../intmd/grms/y/y_all --thread-num 24
gcta64 --bfile ../../intmd/input_data/cd/cd --make-grm --out ../../intmd/grms/cd/cd_all --thread-num 24
gcta64 --bfile ../../intmd/input_data/cl/cl --make-grm --out ../../intmd/grms/cl/cl_all --thread-num 24
gcta64 --bfile ../../intmd/input_data/cy/cy --make-grm --out ../../intmd/grms/cy/cy_all --thread-num 24
gcta64 --bfile ../../intmd/input_data/dl/dl --make-grm --out ../../intmd/grms/dl/dl_all --thread-num 24
gcta64 --bfile ../../intmd/input_data/dy/dy --make-grm --out ../../intmd/grms/dy/dy_all --thread-num 24
gcta64 --bfile ../../intmd/input_data/ly/ly --make-grm --out ../../intmd/grms/ly/ly_all --thread-num 24
gcta64 --bfile ../../intmd/input_data/dly/dly --make-grm --out ../../intmd/grms/dly/dly_all --thread-num 24
gcta64 --bfile ../../intmd/input_data/cdly/cdly --make-grm --out ../../intmd/grms/cdly/cdly_all --thread-num 24
```

# R Scripts

## MakePheno.R {#makepheno}

**Script:**

```{r make-pheno, eval=FALSE}
library(data.table)
library(dplyr)

rm(list=ls())

### Global files

path_ped_plus_intmd <- '../../intmd/pedigree_map/ped_plus_intmd.txt'

ped_plus_intmd <- fread(file = path_ped_plus_intmd,
                     sep = " ",
                     header = F,
                     select = c(1, 7, 15),
                     col.names = c('PigID', 'Line', 'Age'))

### Subgroup 1

path_fam_c <- '../../intmd/input_data/c/c.fam'
path_c_uni_out <- '../../intmd/phenotype/c/c_uni.pheno'

fam_c <- fread(file = path_fam_c,
             sep = " ",
             header = F,
             select = c(1:2),
             col.names = c("FID", "PigID")) %>%
  tibble::rowid_to_column()

pheno_c_uni <- left_join(fam_c, ped_plus_intmd[,c(1, 3)], by="PigID") %>%
  arrange(rowid) %>%
  select(-rowid)
  
fwrite(pheno_c_uni,
       file = path_c_uni_out,
       sep = " ",
       row.names = FALSE,
       col.names = FALSE,
       quote = FALSE,
       na = "NA")  

### Subgroup 2

path_fam_d <- '../../intmd/input_data/d/d.fam'
path_d_uni_out <- '../../intmd/phenotype/d/d_uni.pheno'

fam_d <- fread(file = path_fam_d,
               sep = " ",
               header = F,
               select = c(1:2),
               col.names = c("FID", "PigID")) %>%
  tibble::rowid_to_column()

pheno_d_uni <- left_join(fam_d, ped_plus_intmd[,c(1, 3)], by="PigID") %>%
  arrange(rowid) %>%
  select(-rowid)

fwrite(pheno_d_uni,
       file = path_d_uni_out,
       sep = " ",
       row.names = FALSE,
       col.names = FALSE,
       quote = FALSE,
       na = "NA")  

### Subgroup 3

path_fam_l <- '../../intmd/input_data/l/l.fam'
path_l_uni_out <- '../../intmd/phenotype/l/l_uni.pheno'

fam_l <- fread(file = path_fam_l,
               sep = " ",
               header = F,
               select = c(1:2),
               col.names = c("FID", "PigID")) %>%
  tibble::rowid_to_column()

pheno_l_uni <- left_join(fam_l, ped_plus_intmd[,c(1, 3)], by="PigID") %>%
  arrange(rowid) %>%
  select(-rowid)

fwrite(pheno_l_uni,
       file = path_l_uni_out,
       sep = " ",
       row.names = FALSE,
       col.names = FALSE,
       quote = FALSE,
       na = "NA") 

### Subgroup 4

path_fam_y <- '../../intmd/input_data/y/y.fam'
path_y_uni_out <- '../../intmd/phenotype/y/y_uni.pheno'

fam_y <- fread(file = path_fam_y,
               sep = " ",
               header = F,
               select = c(1:2),
               col.names = c("FID", "PigID")) %>%
  tibble::rowid_to_column()

pheno_y_uni <- left_join(fam_y, ped_plus_intmd[,c(1, 3)], by="PigID") %>%
  arrange(rowid) %>%
  select(-rowid)

fwrite(pheno_y_uni,
       file = path_y_uni_out,
       sep = " ",
       row.names = FALSE,
       col.names = FALSE,
       quote = FALSE,
       na = "NA") 

### Subgroup 5

path_fam_cd <- '../../intmd/input_data/cd/cd.fam'
path_cd_uni_out <- '../../intmd/phenotype/cd/cd_uni.pheno'
path_cd_bi_out <- '../../intmd/phenotype/cd/cd_bi.pheno'

fam_cd <- fread(file = path_fam_cd,
                sep = " ",
                header = F,
                select = c(1:2),
                col.names = c("FID", "PigID")) %>%
  tibble::rowid_to_column()

pheno_cd_uni <- left_join(fam_cd, ped_plus_intmd[,c(1, 3)], by="PigID") %>%
  arrange(rowid) %>%
  select(-rowid)

fwrite(pheno_cd_uni,
       path_cd_uni_out,
       sep = " ",
       row.names = FALSE,
       col.names = FALSE,
       quote = FALSE,
       na = "NA")

pheno_cd_bi <- left_join(fam_cd, ped_plus_intmd, by="PigID") %>%
  rename(Crossbred = Age) %>%
  mutate(Duroc = Crossbred) %>%
  mutate(Crossbred = ifelse(Line == 9006, Crossbred, NA),
         Duroc = ifelse(Line == 1006, Duroc, NA)) %>%
  arrange(rowid) %>%
  select(-c(rowid, Line))

fwrite(pheno_cd_bi,
       path_cd_bi_out,
       sep = " ",
       row.names = FALSE,
       col.names = FALSE,
       quote = FALSE,
       na = "NA")

### Subgroup 6

path_fam_cl <- '../../intmd/input_data/cl/cl.fam'
path_cl_uni_out <- '../../intmd/phenotype/cl/cl_uni.pheno'
path_cl_bi_out <- '../../intmd/phenotype/cl/cl_bi.pheno'

fam_cl <- fread(file = path_fam_cl,
                sep = " ",
                header = F,
                select = c(1:2),
                col.names = c("FID", "PigID")) %>%
  tibble::rowid_to_column()

pheno_cl_uni <- left_join(fam_cl, ped_plus_intmd[,c(1, 3)], by="PigID") %>%
  arrange(rowid) %>%
  select(-rowid)

fwrite(pheno_cl_uni,
       path_cl_uni_out,
       sep = " ",
       row.names = FALSE,
       col.names = FALSE,
       quote = FALSE,
       na = "NA")

pheno_cl_bi <- left_join(fam_cl, ped_plus_intmd, by="PigID") %>%
  rename(Crossbred = Age) %>%
  mutate(Landrace = Crossbred) %>%
  mutate(Crossbred = ifelse(Line == 9006, Crossbred, NA),
         Landrace = ifelse(Line == 10, Landrace, NA)) %>%
  arrange(rowid) %>%
  select(-c(rowid, Line))

fwrite(pheno_cl_bi,
       path_cl_bi_out,
       sep = " ",
       row.names = FALSE,
       col.names = FALSE,
       quote = FALSE,
       na = "NA")

### Subgroup 7

path_fam_cy <- '../../intmd/input_data/cy/cy.fam'
path_cy_uni_out <- '../../intmd/phenotype/cy/cy_uni.pheno'
path_cy_bi_out <- '../../intmd/phenotype/cy/cy_bi.pheno'

fam_cy <- fread(file = path_fam_cy,
                sep = " ",
                header = F,
                select = c(1:2),
                col.names = c("FID", "PigID")) %>%
  tibble::rowid_to_column()

pheno_cy_uni <- left_join(fam_cy, ped_plus_intmd[,c(1, 3)], by="PigID") %>%
  arrange(rowid) %>%
  select(-rowid)

fwrite(pheno_cy_uni,
       path_cy_uni_out,
       sep = " ",
       row.names = FALSE,
       col.names = FALSE,
       quote = FALSE,
       na = "NA")

pheno_cy_bi <- left_join(fam_cy, ped_plus_intmd, by="PigID") %>%
  rename(Crossbred = Age) %>%
  mutate(Yorkshire = Crossbred) %>%
  mutate(Crossbred = ifelse(Line == 9006, Crossbred, NA),
         Yorkshire = ifelse(Line == 11, Yorkshire, NA)) %>%
  arrange(rowid) %>%
  select(-c(rowid, Line))

fwrite(pheno_cy_bi,
       path_cy_bi_out,
       sep = " ",
       row.names = FALSE,
       col.names = FALSE,
       quote = FALSE,
       na = "NA")

### Subgroup 8

path_fam_dl <- '../../intmd/input_data/dl/dl.fam'
path_dl_uni_out <- '../../intmd/phenotype/dl/dl_uni.pheno'
path_dl_bi_out <- '../../intmd/phenotype/dl/dl_bi.pheno'

fam_dl <- fread(file = path_fam_dl,
                sep = " ",
                header = F,
                select = c(1:2),
                col.names = c("FID", "PigID")) %>%
  tibble::rowid_to_column()

pheno_dl_uni <- left_join(fam_dl, ped_plus_intmd[,c(1, 3)], by="PigID") %>%
  arrange(rowid) %>%
  select(-rowid)

fwrite(pheno_dl_uni,
       path_dl_uni_out,
       sep = " ",
       row.names = FALSE,
       col.names = FALSE,
       quote = FALSE,
       na = "NA")

pheno_dl_bi <- left_join(fam_dl, ped_plus_intmd, by="PigID") %>%
  rename(Duroc = Age) %>%
  mutate(Landrace = Duroc) %>%
  mutate(Duroc = ifelse(Line == 1006, Duroc, NA),
         Landrace = ifelse(Line == 10, Landrace, NA)) %>%
  arrange(rowid) %>%
  select(-c(rowid, Line))

fwrite(pheno_dl_bi,
       path_dl_bi_out,
       sep = " ",
       row.names = FALSE,
       col.names = FALSE,
       quote = FALSE,
       na = "NA")

### Subgroup 9

path_fam_dy <- '../../intmd/input_data/dy/dy.fam'
path_dy_uni_out <- '../../intmd/phenotype/dy/dy_uni.pheno'
path_dy_bi_out <- '../../intmd/phenotype/dy/dy_bi.pheno'

fam_dy <- fread(file = path_fam_dy,
                sep = " ",
                header = F,
                select = c(1:2),
                col.names = c("FID", "PigID")) %>%
  tibble::rowid_to_column()

pheno_dy_uni <- left_join(fam_dy, ped_plus_intmd[,c(1, 3)], by="PigID") %>%
  arrange(rowid) %>%
  select(-rowid)

fwrite(pheno_dy_uni,
       path_dy_uni_out,
       sep = " ",
       row.names = FALSE,
       col.names = FALSE,
       quote = FALSE,
       na = "NA")

pheno_dy_bi <- left_join(fam_dy, ped_plus_intmd, by="PigID") %>%
  rename(Duroc = Age) %>%
  mutate(Yorkshire = Duroc) %>%
  mutate(Duroc = ifelse(Line == 1006, Duroc, NA),
         Yorkshire = ifelse(Line == 11, Yorkshire, NA)) %>%
  arrange(rowid) %>%
  select(-c(rowid, Line))

fwrite(pheno_dy_bi,
       path_dy_bi_out,
       sep = " ",
       row.names = FALSE,
       col.names = FALSE,
       quote = FALSE,
       na = "NA")

### Subgroup 10

path_fam_ly <- '../../intmd/input_data/ly/ly.fam'
path_ly_uni_out <- '../../intmd/phenotype/ly/ly_uni.pheno'
path_ly_bi_out <- '../../intmd/phenotype/ly/ly_bi.pheno'

fam_ly <- fread(file = path_fam_ly,
                sep = " ",
                header = F,
                select = c(1:2),
                col.names = c("FID", "PigID")) %>%
  tibble::rowid_to_column()

pheno_ly_uni <- left_join(fam_ly, ped_plus_intmd[,c(1, 3)], by="PigID") %>%
  arrange(rowid) %>%
  select(-rowid)

fwrite(pheno_ly_uni,
       path_ly_uni_out,
       sep = " ",
       row.names = FALSE,
       col.names = FALSE,
       quote = FALSE,
       na = "NA")

pheno_ly_bi <- left_join(fam_ly, ped_plus_intmd, by="PigID") %>%
  rename(Landrace = Age) %>%
  mutate(Yorkshire = Landrace) %>%
  mutate(Landrace = ifelse(Line == 10, Landrace, NA),
         Yorkshire = ifelse(Line == 11, Yorkshire, NA)) %>%
  arrange(rowid) %>%
  select(-c(rowid, Line))

fwrite(pheno_ly_bi,
       path_ly_bi_out,
       sep = " ",
       row.names = FALSE,
       col.names = FALSE,
       quote = FALSE,
       na = "NA")

### Subgroup 11

path_fam_dly <- '../../intmd/input_data/dly/dly.fam'
path_dly_uni_out <- '../../intmd/phenotype/dly/dly_uni.pheno'

fam_dly <- fread(file = path_fam_dly,
               sep = " ",
               header = F,
               select = c(1:2),
               col.names = c("FID", "PigID")) %>%
  tibble::rowid_to_column()

pheno_dly_uni <- left_join(fam_dly, ped_plus_intmd[,c(1, 3)], by="PigID") %>%
  arrange(rowid) %>%
  select(-rowid)

fwrite(pheno_dly_uni,
       file = path_dly_uni_out,
       sep = " ",
       row.names = FALSE,
       col.names = FALSE,
       quote = FALSE,
       na = "NA") 

### Subgroup 12

path_fam_cdly <- '../../intmd/input_data/cdly/cdly.fam'
path_cdly_uni_out <- '../../intmd/phenotype/cdly/cdly_uni.pheno'

fam_cdly <- fread(file = path_fam_cdly,
                 sep = " ",
                 header = F,
                 select = c(1:2),
                 col.names = c("FID", "PigID")) %>%
  tibble::rowid_to_column()

pheno_cdly_uni <- left_join(fam_cdly, ped_plus_intmd[,c(1, 3)], by="PigID") %>%
  arrange(rowid) %>%
  select(-rowid)

fwrite(pheno_cdly_uni,
       file = path_cdly_uni_out,
       sep = " ",
       row.names = FALSE,
       col.names = FALSE,
       quote = FALSE,
       na = "NA") 

```

